src/pages/studyView/resources/FilesAndLinks.tsx

```diff
@@ -0,0 +1,329 @@
+
+export interface IFilesLinksTable {
+    store: StudyViewPageStore;
+}
+
+class FilesLinksTableComponent extends LazyMobXTable<{
+    [id: string]: string;
+}> {}
+
+const RECORD_LIMIT = 500;
+
+async function fetchResourceDataOfPatient(patientIds: Map<string, string>) {
+    const ret: { [key: string]: ResourceData[] } = {};
+    const promises = [];
+    for (let [patientId, studyId] of patientIds) {
+        promises.push(
+            internalClient
+                .getAllResourceDataOfPatientInStudyUsingGET({
+                    studyId: studyId,
+                    patientId: patientId,
+                    projection: 'DETAILED',
+                })
+                .then(data => {
+                    if (patientId in ret) {
+                        ret[patientId].push(...data);
+                    } else {
+                        ret[patientId] = data;
+                    }
+                })
+        );
+    }
+
+    return Promise.all(promises).then(() => ret);
+}
+
+async function fetchFilesLinksData(
+    filters: StudyViewFilter,
+    sampleIdResourceData: { [sampleId: string]: ResourceData[] },
+    searchTerm: string | undefined,
+    sortAttributeId: string | undefined,
+    sortDirection: 'asc' | 'desc' | undefined,
+    recordLimit: number
+) {
+    let sampleClinicalDataResponse = await getAllClinicalDataByStudyViewFilter(
+        filters,
+        searchTerm,
+        sortAttributeId,
+        sortDirection,
+        recordLimit,
+        0
+    );
+
+    const patientIds = new Map<string, string>();
+
+    forEach(sampleClinicalDataResponse.data, (data, uniqueSampleId) => {
+        forEach(data, item => {
+            patientIds.set(item.patientId, item.studyId);
+        });
+    });
+
+    const resourcesPerPatient: { [key: string]: number } = {};
+    const items: { [attributeId: string]: any } = [];
+
+    const itemBuilder = (resourceData: { [key: string]: ResourceData[] }) => {
+        forEach(resourceData, (data: ResourceData[]) => {
+            forEach(data, (resource: ResourceData) => {
+                items.push({
+                    studyId: resource.studyId,
+                    patientId: resource.patientId,
+                    sampleId: resource.sampleId,
+                    resourcesPerPatient: 0,
+                    typeOfResource: resource?.resourceDefinition?.displayName,
+                    description: resource?.resourceDefinition?.description,
+                    url: resource?.url,
+                });
+            });
+
+            if (data && data.length > 0) {
+                if (!(data[0].patientId in resourcesPerPatient))
+                    resourcesPerPatient[data[0].patientId] = 0;
+
+                resourcesPerPatient[data[0].patientId] += data.length;
+            }
+        });
+    };
+
+    itemBuilder(sampleIdResourceData);
+    const resourcesForPatients = await fetchResourceDataOfPatient(patientIds);
+    itemBuilder(resourcesForPatients);
```

Comment:
We can probably combine `sampleIdResourceData` and `resourcesForPatients` and call the `itemBuilder` only once. Also, it's better to rename `itemBuilder` to `buildItemsAndResources` to improve readability.

So, something like this would be better assuming that `buildItemsAndResources` will return those values instead of modifying variables outside the function:

```JavaScript
const {resourcesPerPatient, items} = buildItemsAndResources(resourcesForPatientsAndSamples);
```

Reply:
Okay, I will change it accordingly, however, it will increase time/space complexity.